wurst/math/Matrices.wurst

```diff
@@ -0,0 +1,310 @@
+public function mat3.row(int id) returns vec3
+    switch id
+        case 0
+            return vec3(this.rc00, this.rc01, this.rc02)
+        case 1
+            return vec3(this.rc10, this.rc11, this.rc12)
+        case 2
+            return vec3(this.rc20, this.rc21, this.rc22)
+        default
+            return ZERO3
+        
+public function mat3.determinant() returns real
+    return this.rc00*(this.rc11*this.rc22 - this.rc12*this.rc21)
+         - this.rc01*(this.rc10*this.rc22 - this.rc12*this.rc20)
+         + this.rc02*(this.rc10*this.rc21 - this.rc11*this.rc20)
+
+/** Flips the matrix over its main diagonale.
+Result of transposing for rotation matrix is its inverse matrix. */
+public function mat3.transpose() returns mat3
+    return mat3(
+        this.rc00, this.rc10, this.rc20,
+        this.rc01, this.rc11, this.rc21,
+        this.rc02, this.rc12, this.rc22)
+
+/** Returns a zero-matrix in case there's no iverse matrix.*/
+public function mat3.inverse() returns mat3
+    var d = this.determinant()
+    if d.abs().isBetween(0.998, 1.002)
+        return this.transpose()
+    else if d.abs() > 0.002
+        return 1/d*mat3(
+            this.rc11*this.rc22 - this.rc12*this.rc21,
+            this.rc02*this.rc21 - this.rc01*this.rc22,
+            this.rc01*this.rc12 - this.rc02*this.rc11,
+        
+            this.rc12*this.rc20 - this.rc10*this.rc22,
+            this.rc00*this.rc22 - this.rc20*this.rc02,
+            this.rc02*this.rc10 - this.rc00*this.rc12,
+
+            this.rc10*this.rc21 - this.rc11*this.rc20,
+            this.rc01*this.rc20 - this.rc00*this.rc21,
+            this.rc00*this.rc11 - this.rc01*this.rc10)
+    return ZERO33
+
+/** Extracts Euler-angles (Tait-Bryan) from rotation matrix. Order of result angles is X-Y-Z.*/
+public function mat3.toEuler() returns vec3
+    real yaw
+    real pitch = Asin(this.rc02)
+    real roll
+    var cosY = Cos(pitch)
+    if cosY.abs() > 0.005
+        var cosX = this.rc22/cosY
+        var sinX = -this.rc12/cosY
+        yaw  = Atan2(sinX, cosX)
+        var cosZ =  this.rc00/cosY
+        var sinZ = -this.rc01/cosY
+        roll = Atan2(sinZ, cosZ)
+    else
+        yaw = 0.
+        var cosZ = this.rc11
+        var sinZ = this.rc10
+        roll  = Atan2(sinZ, cosZ) 
+    return vec3(yaw, pitch, roll)
+
+public function mat3.toString() returns string
+    return "Matrix 3x3\n{0} {1} {2}\n{3} {4} {5}\n{6} {7} {8}".format(
+        this.rc00.toString(), this.rc01.toString(), this.rc02.toString(),
+        this.rc10.toString(), this.rc11.toString(), this.rc12.toString(),
+        this.rc20.toString(), this.rc21.toString(), this.rc22.toString())
+
+/** Creates 3x3 rotation matrix from axis and angle. */
+public function vec3.toMatrix(angle angl) returns mat3
+    var v = this.norm()
+    var x = v.x
+    var y = v.y
+    var z = v.z
+    var cos = angl.cos()
+    var sin = angl.sin()
+    return mat3(

```

Comment:
this is a prime example of why I don't like having an empty row `mat3(` followed by un-tabulated entries.
This to me just looks confusing/out of place.
I would suggest filling the first row and then properly aligning the cells, e.g.:

```
mat3(	1 + (1 - cos)*(x*x - 1), 	-z*sin+ (1 - cos)*x*y, 		y*sin+ (1 - cos)*x*z,
	z*sin+ (1 - cos)*x*y, 		1 + (1 - cos)*(y*y - 1), 	-x*sin+ (1 - cos)*y*z,
	-y*sin+ (1 - cos)*x*z, 		x*sin+ (1 - cos)*y*z, 		1 + (1 - cos)*(z*z - 1))
```
This applies to all of these statements above, especially mat2 with short params.

Also these calcs are overly big. You should e.g. save 1 - cos in a var